contracts/liquidityMining/KyberRewardLocker.sol

```diff
@@ -0,0 +1,317 @@
+    address account,
+    uint256 quantity
+  ) external override {
+    lockWithStartTime(token, account, quantity, _blockTimestamp());
+  }
+
+  function lockWithStartTime(
+    IERC20Ext token,
+    address account,
+    uint256 quantity,
+    uint256 startTime
+  ) public override onlyRewardsContract(token) {
+    require(quantity != 0, 'Quantity cannot be zero');
+
+    // transfer token from reward contract to lock contract
+    token.safeTransferFrom(msg.sender, address(this), quantity);
+
+    VestingSchedules storage schedules = accountsVestingSchedules[account][token];
+    uint256 schedulesLength = schedules.length;
+
+    uint256 endTime = startTime.add(lockTime[token]);
+
+    if (schedulesLength == 0) {
+      accountEscrowedBalance[account][token] = quantity;
+    } else {
+      /* Disallow adding new vested XTK earlier than the last one. */
+      require(schedules.data[schedulesLength - 1].endTime < endTime, 'not linear vesting');
+      accountEscrowedBalance[account][token] = accountEscrowedBalance[account][token].add(
+        quantity
+      );
+    }
+    // append to storage, the schedule data
+    schedules.data[schedulesLength] = VestingSchedule({
+      startTime: startTime.toUint64(),
+      endTime: endTime.toUint64(),
+      quantity: quantity.toUint128()
+    });
+    schedules.length = schedulesLength + 1;
+
+    emit VestingEntryCreated(token, account, _blockTimestamp(), quantity);
+  }
+
+  /**
+   * @notice Allow a user to withdraw any XTK in their schedule that have vested.
+   */
+  function vestAll(IERC20Ext token) external returns (uint256) {
+    VestingSchedules storage schedules = accountsVestingSchedules[msg.sender][token];
+    uint256 schedulesLength = schedules.length;
+
+    uint256 totalVesting = 0;
+    uint256 totalSlashing = 0;
+    for (uint256 i = 0; i < schedulesLength; i++) {
+      VestingSchedule memory schedule = schedules.data[i];
+      if (schedule.quantity == 0) {
+        continue;
+      }
+      uint256 vestQuantity = _getVestingQuantity(
+        schedule.quantity,
+        schedule.startTime,
+        schedule.endTime
+      );
+      totalVesting = totalVesting.add(vestQuantity);
+      totalSlashing = totalSlashing.add(schedule.quantity - vestQuantity);
+      // clear data after vesting
+      schedules.data[i].quantity = 0;
+    }
+    require(totalVesting != 0, 'invalid vesting amount');
+    accountEscrowedBalance[msg.sender][token] = accountEscrowedBalance[msg.sender][token].sub(
+      totalVesting.add(totalSlashing)
+    );
+    accountVestedBalance[msg.sender][token] = accountVestedBalance[msg.sender][token].add(
+      totalVesting
+    );
+
+    token.safeTransfer(msg.sender, totalVesting);
+    if (totalSlashing != 0) _slash(token, totalSlashing);
+
+    emit Vested(token, msg.sender, _blockTimestamp(), totalVesting, totalSlashing);
+
+    return totalVesting;
+  }
+
+  function vestAtIndex(IERC20Ext token, uint256[] calldata indexes) external returns (uint256) {
```

Comment:
quite a bit of code duplication with `vestAll`, maybe refactor to internal function

Reply:
Because we also implement slashing so we need this function.
For example:
- user has finish vesting time for schedule 1
- user has not finish vesting time for schedule 2
- he do not want to be slashed, so he can still claim at schedule 1 only

Reply:
Yes, what I'm saying is that the internal logic of `vestAll` and `vestAtIndex` is almost the same; only difference is the iteration indexes, so maybe shift the internal logic to a separate internal function. like `_vest()`, which `vestAll` and `vestAtIndex` will call.

Reply:
change a bit of logic in vestAll. vestAll allows user to vest all ended schedule.

this function supports UI